virtual_rainforest/core/axes.py

```diff
@@ -0,0 +1,396 @@
+        # for that axis must be able to validate the array, otherwise we can skip
+        # validation on this axis and return the input array.
+        matching_dims = validator_dims.intersection(value.dims)
+        if matching_dims:
+
+            # There should be one and only validator that can validate for this axis.
+            validator_found = [v for v in validators if v().can_validate(value, grid)]
+
+            if len(validator_found) == 0:
+                log_and_raise(
+                    f"DataArray uses '{axis}' axis dimension names but does "
+                    f"not match a validator: {','.join(matching_dims)}",
+                    ValueError,
+                )
+
+            if len(validator_found) > 1:
+                log_and_raise(
+                    f"Validators on '{axis}' axis not mutually exclusive", RuntimeError
+                )
+
+            # Get the appropriate Validator class and then use it to update the data
+            # array
+            this_validator = validator_found[0]
+            try:
+                value = this_validator().run_validation(value, grid, **kwargs)
+            except Exception as excep:
+                log_and_raise(str(excep), excep.__class__)
+
+            validation_dict[axis] = this_validator.__name__
+
+        else:
+            validation_dict[axis] = None
+
+    return value, validation_dict
+
+
+class Spat_CellId_Coord_Any(AxisValidator):
+    """Spatial Axis Validator for cell id coordinates on any grid.
+
+    This spatial axis validator applies to a DataArray that has a cell_id dimension with
+    valued coordinates, which should map onto the grid cell ids, allowing for a subset
+    of ids. Because this method simply maps data to grid cells by id, it should apply to
+    _any_ arbitrary grid setup.
+    """
+
+    core_axis = "spatial"
+    dim_names = {"cell_id"}
+
+    def can_validate(  # noqa: D102
+        self, value: DataArray, grid: Grid, **kwargs: Any
+    ) -> bool:
+
+        return self.dim_names.issubset(value.dims) and self.dim_names.issubset(
+            value.coords
+        )
+
+    def run_validation(  # noqa: D102
+        self, value: DataArray, grid: Grid, **kwargs: Any
+    ) -> DataArray:
+
+        da_cell_ids = value["cell_id"].values
+
+        if len(np.unique(da_cell_ids)) != len(da_cell_ids):
+            raise ValueError("The data cell ids contain duplicate values.")
+
+        if not set(grid.cell_id).issubset(da_cell_ids):
+            raise ValueError("The data cell ids are not a superset of grid cell ids.")
+
+        # Now ensure sorting and any subsetting:
+        # https://stackoverflow.com/questions/8251541
+        da_sortorder = np.argsort(da_cell_ids)
+        gridid_pos = np.searchsorted(da_cell_ids[da_sortorder], grid.cell_id)
+        da_indices = da_sortorder[gridid_pos]
+
+        return value.isel(cell_id=da_indices)
+
+
+class Spat_CellId_Dim_Any(AxisValidator):
```

Comment:
Feel the difference between this validator and the one above could be made clearer somewhere

Reply:
I feel like I only half get the distinction